File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/turbomodule/core/interfaces/TurboModuleWithJSIBindings.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.react.turbomodule.core.interfaces

import com.facebook.proguard.annotations.DoNotStrip

/** Implements this interface if a TurboModule needs to install its own JSI bindings. */
@DoNotStrip
public interface TurboModuleWithJSIBindings {
  /**
   * Returns the [BindingsInstallerHolder] that the core will later invoke with an
   * `facebook::jsi::Runtime` instance. The implementation will typically mix with JNI and C++.
   */
  @DoNotStrip public fun getBindingsInstaller(): BindingsInstallerHolder
}

File: packages/react-native/ReactCommon/react/nativemodule/core/ReactCommon/TurboModuleWithJSIBindings.cpp
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

#include "TurboModuleWithJSIBindings.h"

#include <ReactCommon/TurboModule.h>

namespace facebook::react {

/* static */ void TurboModuleWithJSIBindings::installJSIBindings(
    const std::shared_ptr<TurboModule>& cxxModule,
    jsi::Runtime& runtime) {
  if (auto* cxxModuleWithJSIBindings =
          dynamic_cast<TurboModuleWithJSIBindings*>(cxxModule.get())) {
    cxxModuleWithJSIBindings->installJSIBindingsWithRuntime(runtime);
  }
}

} // namespace facebook::react
